### ISSUE TITLE
trial -u leaks memory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3810]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3810|
|Type|release blocker: regression|
|Created|2009-05-01 21:37:52Z|
|Branch|https://github.com/twisted/twisted/tree/stop-thread-pool-3810|

This can be observed by running the following test suite:

```python
from twisted.trial.unittest import TestCase

class Empty(TestCase):
    def test_empty(self):
        pass
```

like this:

```
exarkun@charm:/tmp$ trial -u test_empty.py
```

The cause seems to be some additional ([#3786](https://github.com/twisted/twisted/issues/3786)) threadpool hackery in trial.

Attachments:

* [spike.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/72e/72e9b0837c5a2c2535da9b0e8fe4a0c389b04f51/9a19284f601bcc54af085fea5cb0eecdbb08b03b.patch) (1663 bytes) - added by exarkun on 2009-05-01 21:51:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3810 3810
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_stop_thread_pool_3810 branches/stop-thread-pool-3810
branch_author__exarkun__radix exarkun, radix
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1241213872000000 1241213872000000
changetime__1241617127000000 1241617127000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Attached is a patch which demonstrates the minimum change required to fix the leak.  Clearly it is not actually suitable to be applied as-is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26782]) Branching to 'stop-thread-pool-3810'

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix removed owner|
|-|-|

This is ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

Pretty much every test in the entire suite now produces this error:
```
Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/trial/unittest.py", line 967, in _classCleanUp
    util._Janitor(self, result).postClassCleanup()
  File "/home/glyph/Projects/Twisted/trunk/twisted/trial/util.py", line 150, in postClassCleanup
    self._cleanThreads()
  File "/home/glyph/Projects/Twisted/trunk/twisted/trial/util.py", line 193, in _cleanThreads
    reactor._stopThreadPool()
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/base.py", line 926, in _stopThreadPool
    self.threadpool.stop()
exceptions.AttributeError: 'NoneType' object has no attribute 'stop'
```

they all also say [OK] so ... hrm.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26816]) re-add the check of `reactor.threadpool` being `None` to avoid stopping a threadpool which doesn't exist

refs [#3810](https://github.com/twisted/twisted/issues/3810)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

r26816 fixes the problem.  The cleanup code was trying to destroy the threadpool unconditionally.  If a test method didn't ever trigger the creation of a threadpool, as most don't, this would result in the reported `AttributeError`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

 *
```
+        When L{ReactorBase._stopThreadPool} drops the reactor's direct
+        reference to its internal threadpool and removes the associated startup 
+        and shutdown triggers.
```

I think the sentence is missing something.

 * Pyflakes:
```
twisted/internet/test/test_base.py:8: 'gc' imported but unused
twisted/internet/test/test_base.py:10: 'ref' imported but unused
twisted/internet/test/test_base.py:17: 'ReactorBase' imported but unused
```


Please merge once fixed. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [26858]) Address review feedback

  * fix `test_stopThreadPool` docstring
  * remove unused imports from `test_base.py`

refs [#3810](https://github.com/twisted/twisted/issues/3810)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26860]) Update copyright date

refs [#3810](https://github.com/twisted/twisted/issues/3810)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [26861]) Merge stop-thread-pool-3810

Author: radix, exarkun
Reviewer: glyph, therve
Fixes: [#3810](https://github.com/twisted/twisted/issues/3810)

Change the interaction between trial and the reactor in the area of threadpool
cleanup.  Move more responsibility into the reactor so that trial is doing (very)
slightly less to objects it isn't really responsible for.

Specifically, make the reactor's threadpool cleanup method also clean up the
startup trigger it adds to itself, if necessary, and make trial only call the
reactor's threadpool cleanup method when it wants to clean up threads.

Also add test coverage for the reactor's threadpool cleanup code for each of
the various states it is possible for the reactor to be in.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
